### PR TITLE
Generate final sample of OJO

### PR DESCRIPTION
**Pull request by india-kerle**
---

# Description

This PR generates a final, random sample of job ads. 

Please include a summary of the changes.

Fixes # (issue)

#95 

# Instructions for Reviewer

You can use getters to test the size of the sample here:

```
from dap_prinz_green_jobs.getters.ojo_getters import(
    get_large_ojo_sample,
    get_large_ojo_job_title_sample,
    get_large_ojo_location_sample,
    get_large_ojo_salaries_sample,
    get_large_ojo_skills_sample
)

ojo_data = get_large_ojo_sample()
ojo_job_titles = get_large_ojo_job_title_sample()
ojo_locations = get_large_ojo_location_sample()
ojo_salaries = get_large_ojo_salaries_sample()
ojo_skills = get_large_ojo_skills_sample()
```

This should be our final large sample.

Please pay special attention to ...

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/sample_ojo.py [hunk 1]
```diff
@@ -69,33 +75,87 @@ def filter_data(data, filter_ids_set, id_col_name="id"):
...
+        job_title_locations.groupby(["soc4_code_itl2_code"]).size()
+        / job_title_locations.shape[0]
+    )
+    job_title_locations[
+        "soc4_code_itl2_code"
+    ] = job_title_locations.soc4_code_itl2_code.map(soc4_itl2_weights)
+
+    large_sample_df = job_title_locations.sample(
+        n=desired_sample_size, weights="soc4_code_itl2_code", random_state=random_seed
+    )
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/sample_ojo.py]: this is how i sampled by soc4 codes and itl2 codes i.e. passed the combinations as weights - this is what i'd focus on in the review!

**Comment by india-kerle** [comment]
FYI I did a bit of analysis in terms of deduplication and the over/random sampling approaches (in `sampling_analysis.ipynb`):

For **deduplication**:

- there are **3618729** unique, deduplicated ids
- there are **7883041** unique ids from the full OJO sample job titles dataset
- **45.91%** of job ads are duplicates

On sampling:
- all ITL 2/3 codes are represented in both the over and randomly sampled job ads (with the exception of 1 itl3 code in the oversampled dataset)
For the oversampled dataset...
- 2.0% of itl 2 codes have less than 100 job ads
- 56.99999999999999% of itl 3 codes have less than 50 job ads
- 71.0% of itl 3 codes have less than 100 job ads

For the randomly sampled dataset...
- 0.0% of itl 2 codes have less than 100 job ads
- 2.0% of itl 3 codes have less than 50 job ads
- 7.000000000000001% of itl 3 codes have less than 100 job ads

I also looked into % of job ads per soc4 code per itl 2 regions + top regions more generally (see graph attached). 

visualization

**Comment by india-kerle** [comment]
A few thoughts...


- the **weighted sample** is quite skewed at the itl3 level i.e. most regions have fewer than 100 job ads and half have fewer than 50 job ads.
- the deduplication method does get rid of almost half of the job adverts: I wonder if its worth re-running the script as I know jack's been collecting job ads? I don't think that will change the results drastically but could be good to be updated
- we would still need to oversample in a few regions if we move forward with a totally random sample

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py [hunk 2]
```diff
@@ -37,9 +40,7 @@
 
     # Merge what's needed for the deduplication
     job_adverts = adverts_ojd_daps_extract[["id", "job_location_raw", "created"]]
-    job_adverts.loc[:, "description_hash"] = job_adverts["id"].apply(
-        lambda x: hash_dict.get(str(x))
-    )
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py:42]: for some reason the hash dict didn't map onto the ids so i've updated this

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/get_soc_codes.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
+"""
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/get_soc_codes.py:1]: this script isn't being used any more !

**Comment by india-kerle** [comment]
@lizgzil this is ready for review now! I'm just randomly sampling now

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/get_soc_codes.py (hunk 3 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/get_soc_codes.py:1]: maybe we could keep it in for a rainy day

Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/get_soc_codes.py:1]: I think the output might be interesting anyway, e.g. if we ever include analysis on how our sample fits into the full dataset. We could even include it on the occupation page for the green jobs explorer - e.g. "this occupation makes up 2% of our full dataset"

**Review by lizgzil — APPROVED** [review]
beautiful! All looks good to me

**Pull request merged**